# Capstone Assignment 20.1: Initial Report and Exploratory Data Analysis (EDA)

Nathan Oyama

## 1 &emsp; Planning the project

There are three data sets to accomplish this project. 

* Kaggle: "Percent Sunshine by US City". kaggle.com/datasets/thedevastator/annual-percent-of-possible-sunshine-by-us-city.

* US Geological Survey: "The United States Large-Scale Solar Photovoltaic Database (USPVDB)". US Department of the Interior. energy.usgs.gov/uspvdb/data.

* landvalue: "ZHVI 3-Bedroom Time Series($) - City". landvalue.com/research/data.

Then take the following steps _for every dataset_:

1. From the data set which is in CSV format, create a pandas DataFrame object.
1. Analyze every DataFrame and identify which columns to use for this project.
1. Format the DataFrames before merge them.

Finally, merge the three DataFrames into one.

## 2 &emsp; Analyzing Data Sets

Analyze those three data sets.

In [180]:
import pandas as pd
import re

pd.options.mode.copy_on_write = True

### 2.1 Analyzing Data Set 1: "Percent Sunshine by US City"

In [181]:
df_sunshine_original = pd.read_csv(
    './data/Average Percent of Possible Sunshine by US City.csv'
    )

print(df_sunshine_original.shape)
print(df_sunshine_original.head())

(317, 16)
   index           CITY JAN FEB MAR APR MAY JUN JUL AUG SEP OCT NOV DEC ANN  \
0      0  BIRMINGHAM,AL  46  53  57  65  65  67  59  62  59  66  55  49  58   
1      1  MONTGOMERY,AL  47  55  58  64  63  64  61  61  59  63  55  49  58   
2      2   ANCHORAGE,AK  43  46  51  50  51  46  43  43  41  36  35  33  43   
3      3      JUNEAU,AK  39  35  38  42  44  37  33  35  27  21  26  21  33   
4      4        NOME,AK  38  56  54  52  52  43  39  34  38  35  30  36  42   

   Unnamed: 14  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  


You can use all columns except "index" and "Unnamed: 14". The "CITY" column includes the name of the city in all uppercase, followed by a comma (",") and the state abbreviation.

In [182]:
del df_sunshine_original

### 2.2 &emsp; Analyzing Data Set 2: "The US Large-Scale Solar Photovoltaic Database (USPVDB)"

In [183]:
df_photovoltanic_original = pd.read_csv(
    './data/uspvdb_v3_0_20250430.csv'
    )

print(df_photovoltanic_original.shape)

(5712, 26)


In [184]:
print(df_photovoltanic_original.iloc[:,:13].head())

   case_id multi_poly  eia_id p_state           p_county       ylat  \
0   406374     single   66887      AK  Matanuska-Susitna  61.587349   
1   405016      multi    6304      AK   Northwest Arctic  66.838470   
2   401476      multi   60058      AL         Lauderdale  34.833809   
3   401865      multi   60679      AL               Dale  31.331732   
4   401866      multi   60680      AL            Calhoun  33.626301   

        xlong   p_area  p_img_date  p_dig_conf                   p_name  \
0 -149.789413   172005    20240814           4            Houston Solar   
1 -162.553146     8740    20240719           4          Kotzebue Hybrid   
2  -87.838394  1735134    20220212           4    River Bend Solar, LLC   
3  -85.729469   187820    20220609           4  Fort Rucker Solar Array   
4  -85.940590    39717    20210814           4         ANAD Solar Array   

   p_year p_pwr_reg  
0    2023        AK  
1    2020       NaN  
2    2016       TVA  
3    2017      SOCO  
4    2017   

In [185]:
print(df_photovoltanic_original.iloc[:,13:26].head())

  p_tech_pri p_tech_sec p_sys_type       p_axis  p_azimuth  p_tilt  p_battery  \
0         PV        NaN     ground   fixed-tilt      180.0    40.0        NaN   
1         PV        NaN     ground  single-axis      156.0    40.0  batteries   
2         PV       c-si     ground  single-axis      270.0    17.0        NaN   
3         PV  thin-film     ground  single-axis      188.0    20.0        NaN   
4         PV  thin-film     ground   fixed-tilt      180.0    20.0        NaN   

   p_cap_ac  p_cap_dc      p_type       p_agrivolt p_comm  p_zscore  
0       6.0       8.4  greenfield             crop    NaN -0.457675  
1       1.7       3.4  greenfield  non-agrivoltaic    NaN  5.617232  
2      75.0     100.2  greenfield  non-agrivoltaic    NaN -0.298527  
3      10.6      12.7  greenfield  non-agrivoltaic    NaN -0.122265  
4       7.4       9.7   superfund  non-agrivoltaic    NaN  3.031619  


In [186]:
del df_photovoltanic_original

### 2.3 &emsp; Analyzing Data Set 3: "ZHVI 3-Bedroom Time Series($) - City"

In [187]:
df_landvalue_original = pd.read_csv(
    './data/City_zhvi_bdrmcnt_3_uc_sfrcondo_tier_0.33_0.67_sm_sa_month.csv.zip',
    compression='zip'
    )

df_landvalue_original.head()

RegionID  SizeRank   RegionName RegionType StateName State  \
0      6181         0     New York       city        NY    NY   
1     12447         1  Los Angeles       city        CA    CA   
2     39051         2      Houston       city        TX    TX   
3     17426         3      Chicago       city        IL    IL   
4      6915         4  San Antonio       city        TX    TX   

                                   Metro          CountyName     2000-01-31  \
0  New York-Newark-Jersey City, NY-NJ-PA       Queens County  260254.899991   
1     Los Angeles-Long Beach-Anaheim, CA  Los Angeles County  225748.889853   
2   Houston-The Woodlands-Sugar Land, TX       Harris County   94776.860067   
3     Chicago-Naperville-Elgin, IL-IN-WI         Cook County  129815.966452   
4          San Antonio-New Braunfels, TX        Bexar County   91255.267745   

      2000-02-29     2000-03-31     2000-04-30     2000-05-31     2000-06-30  \
0  261634.936152  263035.285487  265755.628388  268537.262690  271540.803319   
1  226055.608837  227045.607934  229008.591848  231482.858553  233877.634570   
2   94776.153916   94666.014300   94674.385216   94704.008600   94896.016105   
3  129799.128540  130057.282247  130769.827932  131729.662557  132773.661649   
4   91346.835579   91400.169521   91507.390185   91097.854708   90546.511330   

      2000-07-31     2000-08-31     2000-09-30     2000-10-31     2000-11-30  \
0  274684.685195  277577.034679  280225.595389  283007.236196  286304.212798   
1  236308.780654  238809.005183  241114.916062  242973.009387  244651.981084   
2   95081.733616   95372.460970   95758.619977   96123.042959   96637.943931   
3  133733.957119  134929.372959  136228.373549  137756.960004  139513.465895   
4   89941.934818   89832.043040   89851.403034   89898.629113   89885.726028   

      2000-12-31     2001-01-31     2001-02-28     2001-03-31     2001-04-30  \
0  289526.956545  292654.837159  295515.354541  298638.996994  301808.051312   
1  246244.549109  247280.934728  248273.002967  249682.776906  252000.133987   
2   97180.122996   97462.032292   97483.685445   97172.735866   97020.213630   
3  141223.685121  142639.244918  143511.345882  144337.753346  145283.810906   
4   89847.540662   89690.816430   89453.201635   89167.867345   88954.558780   

      2001-05-31     2001-06-30     2001-07-31     2001-08-31     2001-09-30  \
0  304718.323281  307522.118851  310207.310883  313169.182122  315895.295344   
1  254508.404555  256970.470396  259538.936397  262199.221867  264634.049487   
2   97028.023045   97179.873838   97362.562853   97384.242163   97527.660212   
3  146365.409919  147441.229396  148343.429364  149484.760732  150824.471291   
4   88871.135262   88862.562161   88906.202149   88936.277317   88913.870312   

      2001-10-31     2001-11-30     2001-12-31     2002-01-31     2002-02-28  \
0  318551.360488  321587.071324  325155.741965  329092.979419  332385.088893   
1  267023.209084  269188.313235  271393.248285  273016.311806  274845.437224   
2   97737.720217   98007.576422   98199.221917   98330.786171   98484.692803   
3  152470.334050  154121.480370  155517.616153  156712.858766  157629.139473   
4   88875.895436   88802.930547   88755.572967   88682.848073   88656.952948   

      2002-03-31     2002-04-30     2002-05-31     2002-06-30     2002-07-31  \
0  333194.788341  332854.557386  332395.370390  334067.937165  336510.274755   
1  277281.862831  280745.362308  284766.932549  288854.895183  293730.463622   
2   98862.699241   99420.769557  100012.617386  100525.407868  100996.593884   
3  158656.079683  159559.466842  160503.810361  161269.184390  162254.927682   
4   88742.793627   89003.403005   89363.234084   89717.247691   90050.699419   

      2002-08-31     2002-09-30     2002-10-31     2002-11-30     2002-12-31  \
0  339775.608296  343119.158839  346568.801946  349275.695088  351788.680626   
1  298875.474066  304274.727278  309219.908578  314076.343917  318741.526519   
2  101538.7

In [188]:
df_landvalue_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15711 entries, 0 to 15710
Columns: 313 entries, RegionID to 2025-05-31
dtypes: float64(305), int64(2), object(6)
memory usage: 37.5+ MB


In [189]:
df_landvalue_original.columns

Index(['RegionID', 'SizeRank', 'RegionName', 'RegionType', 'StateName',
       'State', 'Metro', 'CountyName', '2000-01-31', '2000-02-29',
       ...
       '2024-08-31', '2024-09-30', '2024-10-31', '2024-11-30', '2024-12-31',
       '2025-01-31', '2025-02-28', '2025-03-31', '2025-04-30', '2025-05-31'],
      dtype='object', length=313)

In [190]:
del df_landvalue_original

### 2.4 &emsp; Planning the merged Data Frame


| Column              | Example          | Data Sets                           |
| :------------------ | :--------------- | :---------------------------------- |
| State               | AL               | Sunshine, Photovoltanic, Land Value |
| County              | DALE             | Photovoltanic, Land Value           |
| City                | BIRMINGAM        | Sunshine, Land Value                |
| Longitude           | -149.789413.     | Photovoltanic                       |
| Latitude            | 61.587349        | Sunshine, Photovoltanic, Land Value |
| Sunshine (Jan ... Dec)  | 58           | Sunshine                            |
| DC                  | 6.0              | Photovoltanic                       |
| AC                  | 8.4              | Photovoltanic                       |
| Current             | 14.4             | (DC + AC)                           |
| Land Value          | 840048.900963529 | Land Value                          |

Some cities that are listed in the Land Value data set and the Sunshine data set do not have solar power plants. In later steps, you predict the current in cities that do not have solar power plants and tell which cities you should build solar power plants.

## 3 &emsp; Extracting Data Sets

### 3.1 &emsp; Dataset 1: Sunshine

In [191]:
df_sunshine_original = pd.read_csv(
    './data/Average Percent of Possible Sunshine by US City.csv'
    )

Check the `CITY` column:

In [192]:
print(df_sunshine_original[['CITY']].value_counts(ascending=True))

CITY                  
CITY                      1
ABERDEEN,SD               2
NOME,AK                   2
NORFOLK,VA                2
NORTH PLATTE,NE           2
                         ..
GRAND RAPIDS,MI           2
GREAT FALLS,MT            2
GREEN BAY,WI              2
FLAGSTAFF,AZ              2
YAP- W CAROLINE IS.,PC    2
Name: count, Length: 159, dtype: int64


There is one invalid entry `"CITY"`, and every other city has exact two entries. Check the row where the `CITY` column is `CITY`:

In [193]:
print(df_sunshine_original[df_sunshine_original['CITY'] == 'CITY'])

     index  CITY  JAN  FEB  MAR  APR  MAY  JUN  JUL  AUG  SEP  OCT  NOV  DEC  \
158    158  CITY  JAN  FEB  MAR  APR  MAY  JUN  JUL  AUG  SEP  OCT  NOV  DEC   

     ANN  Unnamed: 14  
158  ANN          NaN  


In [194]:
print(df_sunshine_original.sort_values(by=['CITY','index']).head(10))

     index            CITY  JAN  FEB  MAR  APR  MAY  JUN JUL AUG SEP OCT NOV  \
115    115     ABERDEEN,SD  NaN   54   58   63   65   66  74  78  68  48  21   
315    315     ABERDEEN,SD  NaN   54   58   63   65   66  74  78  68  48  21   
123    123      ABILENE,TX   63   66   70   71   71   77  80  75  69  68  64   
182    182      ABILENE,TX   63   66   70   71   71   77  80  75  69  68  64   
86      86       ALBANY,NY   46   52   51   55   53   55  62  58  54  46  33   
287    287       ALBANY,NY   46   52   51   55   53   55  62  58  54  46  33   
84      84  ALBUQUERQUE,NM   73   73   73   78   80   82  76  76  77  80  75   
169    169  ALBUQUERQUE,NM   73   73   73   78   80   82  76  76  77  80  75   
108    108    ALLENTOWN,PA  NaN  NaN  NaN  NaN  NaN  NaN  90  93  82  52  47   
314    314    ALLENTOWN,PA  NaN  NaN  NaN  NaN  NaN  NaN  90  93  82  52  47   

     DEC  ANN  Unnamed: 14  
115  NaN  NaN          NaN  
315  NaN  NaN          NaN  
123   65   69          NaN  
182

In this data set, the row where `'index'` is 158 does not include appopriate values thus you can remvoe it.

See the `"index"` field of every pairs of rows for every city; one of those rows have the value of less than 158 and the other rows have the value of greater than 158. Also all other values such as `"ANN"` are the same.

Assume that this data set includes two tables with the same rows in different orders. The first table spans between line 1 and line 157 in the original CSV file, and the second table spans between line 158 and line 318. You only need the first one.

Construct a new DataFrame that is based on the original data set for the sunshine information with the following changes:

* Include all rows of the first 157 records.
* Trim off the unnecessary column `"Unnamed: 14"`.
* Split the "`CITY`" entries by the city names and the state abbreviations.

In [195]:
df_sunshine = df_sunshine_original\
    .iloc[1:157]\
    .set_index(['index'])

del df_sunshine['Unnamed: 14']

df_sunshine[['CITY', 'STATE']] = \
    df_sunshine['CITY'].str.split(',', expand=True)

# del df_sunshine['CITY']

print(df_sunshine.head())

             CITY JAN FEB MAR APR MAY JUN JUL AUG SEP OCT NOV DEC ANN STATE
index                                                                      
1      MONTGOMERY  47  55  58  64  63  64  61  61  59  63  55  49  58    AL
2       ANCHORAGE  43  46  51  50  51  46  43  43  41  36  35  33  43    AK
3          JUNEAU  39  35  38  42  44  37  33  35  27  21  26  21  33    AK
4            NOME  38  56  54  52  52  43  39  34  38  35  30  36  42    AK
5       FLAGSTAFF  71  73  72  82  83  88  74  75  79  77  72  76  76    AZ


In [196]:
del df_sunshine_original

### 3.2 &emsp; Data Set 2: Photovoltanic

In [197]:
df_photovoltanic_original = pd.read_csv(
    './data/uspvdb_v3_0_20250430.csv'
    )

df_photovoltanic = pd.DataFrame()

df_photovoltanic = df_photovoltanic_original[[
    'case_id', 'p_county', 'p_state', 'xlong', 'ylat', 'p_cap_ac', 'p_cap_dc'
    ]].set_index('case_id')

df_photovoltanic.rename(
    columns={
        'p_county': 'County',
        'p_state':  'State',
        'xlong':    'Longitude',
        'ylat':     'Latitude',
        'p_cap_ac': 'AC',
        'p_cap_dc': 'DC'
        },
    inplace=True
    )

df_photovoltanic['Current'] = df_photovoltanic['AC'] + df_photovoltanic['DC'] 

print(df_photovoltanic.head())

                    County State   Longitude   Latitude    AC     DC  Current
case_id                                                                      
406374   Matanuska-Susitna    AK -149.789413  61.587349   6.0    8.4     14.4
405016    Northwest Arctic    AK -162.553146  66.838470   1.7    3.4      5.1
401476          Lauderdale    AL  -87.838394  34.833809  75.0  100.2    175.2
401865                Dale    AL  -85.729469  31.331732  10.6   12.7     23.3
401866             Calhoun    AL  -85.940590  33.626301   7.4    9.7     17.1


### 3.3 &emsp; Dataset 3: Land Values

In this original data set, there are many columns of historical prices of the average home values of 3-bedroom houses but you only need the latest values: `"2025-05-31"`.

See the first few rows of the original data set while excluding all the other columns for historical home values:

In [198]:
df_landvalue_original = pd.read_csv(
    './data/City_zhvi_bdrmcnt_3_uc_sfrcondo_tier_0.33_0.67_sm_sa_month.csv.zip',
    compression='zip'
    )

df_landvalue = pd.DataFrame()

df_landvalue = df_landvalue_original[[
    'RegionID', 'SizeRank', 'State', 'RegionName', 'CountyName', '2025-05-31'
    ]].set_index('RegionID')

df_landvalue['County']\
    = df_landvalue['CountyName']\
        .str.replace(r'\s* County$', '', regex=True)\
        .str.upper()
del df_landvalue['CountyName']

df_landvalue['City'] \
    = df_landvalue['RegionName']\
        .str.upper()
del df_landvalue['RegionName']

df_landvalue['Land Value'] \
    = df_landvalue['2025-05-31']
del df_landvalue['2025-05-31']

del df_landvalue_original
print(df_landvalue.head())

          SizeRank State       County         City     Land Value
RegionID                                                         
6181             0    NY       QUEENS     NEW YORK  840048.900964
12447            1    CA  LOS ANGELES  LOS ANGELES  964249.977821
39051            2    TX       HARRIS      HOUSTON  253134.060059
17426            3    IL         COOK      CHICAGO  336756.496352
6915             4    TX        BEXAR  SAN ANTONIO  235986.092899


### 1.4 &emsp; Combining Three Data Sets into One DataFrame

| Column | Land Value  | Solar   | Sunshine | (Example 1)  | (Example 2)  |
| :----- | :---------: | :-----: | :------: | :----------- | :----------- |
| City   | &#9679;     | -       | &#9679;  | Franklin     | Franklin     |
| County | &#9679;     | &#9679; | -        | Williamson   | Williamson   |
| State  | &#9679;     | &#9679; | &#9679;  | Tennessee    | Texas        |


Note that in the United States, city names are unique only within the same state. Some cities share the same name and even the same county name while those are located in different states.